iceoryx_hoofs/test/moduletests/test_posix_named_semaphore.cpp

```diff
@@ -0,0 +1,263 @@
+        EXPECT_TRUE(isSemaphoreValueEqualTo(*sut2, INITIAL_VALUE));
+    }
+
+    // if the dtor of sut2 unlinks the semaphore we should be unable to open it again
+    {
+        iox::cxx::optional<NamedSemaphore> sut2;
+        ASSERT_FALSE(NamedSemaphoreBuilder()
+                         .name(sutName)
+                         .initialValue(0U)
+                         .openMode(OpenMode::OPEN_OR_CREATE)
+                         .create(sut2)
+                         .has_error());
+        // value should be INITIAL_VALUE since we opened an existing semaphore
+        EXPECT_TRUE(isSemaphoreValueEqualTo(*sut2, INITIAL_VALUE));
+    }
+
+    // verify that the original semaphore which has the ownership still works
+    EXPECT_TRUE(isSemaphoreValueEqualTo(*sut, INITIAL_VALUE));
+}
+
+TEST_F(NamedSemaphoreTest, OpenOrCreateRemovesSemaphoreWhenItHasTheOwnership)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "06252f80-42ee-4d81-b25a-5f3b196b9d7c");
+
+    ASSERT_FALSE(NamedSemaphoreBuilder()
+                     .name(sutName)
+                     .openMode(OpenMode::OPEN_OR_CREATE)
+                     .permissions(sutPermission)
+                     .create(sut)
+                     .has_error());
+
+    sut.reset();
+
+    // should fail since the previous sut was deleted and had the ownership
+    auto result = NamedSemaphoreBuilder().name(sutName).initialValue(0U).openMode(OpenMode::OPEN_EXISTING).create(sut);
+    ASSERT_TRUE(result.has_error());
+    EXPECT_THAT(result.get_error(), Eq(SemaphoreError::NO_SEMAPHORE_WITH_THAT_NAME_EXISTS));
+}
+
+// Windows does not support this since the named semaphore is automatically deleted
+// as soon as the last handle was closed with CloseHandle
+#if !defined(_WIN32)
+TEST_F(NamedSemaphoreTest, PurgeAndCreateCreatesNewSemaphore)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "4b9d055c-b980-4ed6-941b-f93b3810e0e5");
+
+    ASSERT_FALSE(NamedSemaphoreBuilder()
+                     .name(sutName)
+                     .openMode(OpenMode::PURGE_AND_CREATE)
+                     .permissions(sutPermission)
+                     .initialValue(0U)
+                     .create(sut)
+                     .has_error());
+
+    iox::cxx::optional<NamedSemaphore> sut2;
+    constexpr uint32_t INITIAL_VALUE = 97U;
+    ASSERT_FALSE(NamedSemaphoreBuilder()
+                     .name(sutName)
+                     .openMode(OpenMode::PURGE_AND_CREATE)
+                     .permissions(sutPermission)
+                     .initialValue(INITIAL_VALUE)
+                     .create(sut2)
+                     .has_error());
+
+    // the newly created semaphore should have the INITIAL_VALUE
+    EXPECT_TRUE(isSemaphoreValueEqualTo(*sut2, INITIAL_VALUE));
+
+    // the original semaphore cannot be opened anymore but the current open handle should
+    // still be valid
+    EXPECT_TRUE(isSemaphoreValueEqualTo(*sut, 0U));
+}
+#endif
+} // namespace
```

Comment:
In both scenarios the behavior is the same.
When the owning semaphore goes out of scope it will call `sem_close` to close its file handle and `sem_unlink` to remove the semaphore so that no one is able to open it again.
Every semaphore which was opened before the `sem_unlink` call is still valid. The OS cleans up the semaphore resource when:
1. Every user has called `sem_close` on its file handle
2. `sem_unlink` was called.

I added a test to verify this behavior.